Перенос состояния в Redux. Изменение порядка ингредиентов с React DND

EldenLordCommander opened the pull request:
Добрый день. Я сделал задачи указанные в ТЗ, вроде как более менее разобрался как работает Redux, но с трудом. Некоторые операции сделал как понимал, по аналогии с тренажером, чтобы реализовать необходимые функции:

- подключил все библиотеки, указанные в ТЗ
- подключил dev tools, в котором отображаются вызванные экшены, состояние хранилища и изменения
- перешел с контекста на хранилище Redux, с которым работаю через useDispatch и useSelector, с thunk
- описал экшены и редьюсеры для получения ингридиентов из АПИ, получения заказа из АПИ, добавления/удаление элементов в конструктор путем перетаскивания и модалку, которая получает заказ/ингредиент
- добавил перетаскивание и сортировку элементов через DND
- добавил прокрутку с выделением текущего таба списка ингредиентов и навигацию по табам
- вызовы АПИ реализовал с усилителями

Буду рад, если проверите текущий вариант проект и подскажите, в правильном ли направлении движусь и правильно ли понимаю, как нужно работать с redux и с его помощью реализовывать функциональность.

EldenLordCommander committed:
Перенос состояния в Redux. Изменение порядка ингредиентов с React DND

- подключил все библиотеки
- подключил dev tools
- перешел с контекста на хранилище Redux
- описал экшены и редьюсеры для получения ингридиентов из АПИ, получения заказа из АПИ, добавления/удаление элементов в конструктор и модалку
- добавил перетаскивание и сортировку элементов через DND

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
 
 import { IngredientContext } from '../../services/burger-context.js'
 
 const OrderContext = React.createContext(null);
```
Нужно исправить: стоит избавить от неиспользуемой константы и контекста

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
-    const firstIngredient = data[0];
-    const leftIngredients = data.slice(1, data.length - 1);
-    const [openModal, setModal] = useState(false);
+    const data = useSelector((store) => store);
```
Можно лучше: data нигде не используется в коде (лишняя строчка, лучше убирать подобные)

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
+    //console.log(data);
+    const dispatch = useDispatch();
+
+    const dataConstructor = useSelector((store) => store.burgerConstructor.dataConstructor);
```
Отлично: правильно реализована работа с хуком useSelector

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
+    const dispatch = useDispatch();
+
+    const dataConstructor = useSelector((store) => store.burgerConstructor.dataConstructor);
+    const firstIngredient = dataConstructor.bun;
```
Можно лучше: firstIngredient и leftIngredients, можно было не объявлять, а получить свойства bun и components выше,
 используя деструктуризацию объекта dataConstructor (const dataConstructor =  эквивалентно const {bun, components} =)

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
+    // const initialPriceState = { price: firstIngredient.price * 2 };
+    // const [priceState, priceDispatcher] = useReducer(reducer, initialPriceState);
 
     const [orderData, setOrder] = useState({
```
Можно лучше: так же избыточное состояние (нигде не используется)

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,134 +1,181 @@
         }
     }
 
+  const moveCard = useCallback((dragIndex, hoverIndex) => {
```
Отлично: хороший кейс для применения хука useCallback

vadmas commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -140,4 +187,5 @@ ConstructorElement.propTypes = {
     thumbnail: PropTypes.string.isRequired
```
Нужно исправить: описывать типы пропсов компонента следует в файле одноименного компонента.
Так же описывать типы компонента ConstructorElement - не стоит, т.к. это экспортируемый из внешней библиотеки компонент

vadmas commented on src/components/ingredient-details/ingredient-details.js:
```diff
@@ -2,8 +2,13 @@ import React, { useState } from 'react';
+function IngredientDetails({ item }) {
+
+    //const ingredient = useSelector((store) => store.ingredient.modalItem);
+    const ingredient = item;
```
Можно лучше: все верно, т.к. нет вложенности между данным компонентом и родительским в данном случае лучше, то уместнее передать параметр в качестве пропса. Однако не стоит переприсваивать тоже самое значение в новую переменную ingredient

vadmas commented on src/components/ingredient-details/ingredient-details.js:
```diff
@@ -41,7 +46,7 @@ function IngredientDetails({ ingredient }) {
 }
 
 IngredientDetails.propTypes = {
```
Отлично: хорошо описаны типы пропсов

vadmas commented on src/components/ingredient-item/ingredient-item.js:
```diff
@@ -0,0 +1,51 @@
+
+export default function IngredientItem({ item }) {
+
+    const data = useSelector((store) => store.ingredient.data);
```
Можно лучше: data - неиспользуемая переменная, стоит убрать

vadmas commented on src/services/actions/ingredients-actions.js:
```diff
@@ -0,0 +1,70 @@
+export const GET_DATA_SUCCESS = 'GET_DATA_SUCCESS';
+export const GET_DATA_FAILED = 'GET_DATA_FAILED';
+
+// export const GET_ORDER_REQUEST = 'GET_ORDER_REQUEST';
```
Можно лучше: так как данные экшены не участвуют в работе редъюсеров то от них лучше избавиться

vadmas commented on src/services/actions/ingredients-actions.js:
```diff
@@ -0,0 +1,70 @@
+
+
+
+export function getData() {
```
Отлично: правильно реализована работа с санками (redux-thunk)

vadmas commented on src/services/actions/order-action.js:
```diff
@@ -0,0 +1,30 @@
+                    data: res
+                })
+            } else {
+                dispatch({
```
Можно лучше: обработку ошибок лучше осуществлять в конструкции catch
Для этого в данном блоке кода можно передать ошибку в catch: throw new Error(res.errorMessage)

vadmas commented on src/services/actions/order-action.js:
```diff
@@ -0,0 +1,30 @@
+                    type: GET_ORDER_FAILED
+                })
+            }
+        }).catch(err => {
```
Можно лучше: уместнее отражать реальную ошибку запроса, и передавать ее в экшен

vadmas commented on src/services/store.tsx:
```diff
@@ -0,0 +1,12 @@
+ import { configureStore } from '@reduxjs/toolkit'
+
+// import postsReducer, { postsSlice } from './reducers/ingredientsReducers'
```
Можно лучше: данный метод объявления стора в классической реализации работы с redux
(то как это реализовано в данной работе) - не уместен, стоит убрать

vadmas commented on src/index.tsx:
```diff
@@ -4,12 +4,39 @@ import './index.css';
+    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
+  }
+}
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
```
Отлично: правильно подключено расширение REDUX_DEVTOOLS

vadmas commented on src/index.tsx:
```diff
@@ -4,12 +4,39 @@ import './index.css';
+
+const store = createStore(rootReducer, enhancer); 
+
+// const store = configureStore({
```
Можно лучше: данный метод объявления стора актуален для сторов, описаннных с использованием библиотеки @reduxjs/toolkit
Следовательно в данной реализации подобное объявление было бы лишним.

vadmas reviewed (commented):
Хорошая работа.
Приложение начинает приобретать готовый вид! Видно, что студент старательно относится к выполнению заданий.
В положительную сторону хочется отметить:
- Очень ответственное отношение к заданию, и даже в github студент оставил краткий тезис о данной работе
- Подключено расширение REDUX_DEVTOOLS
- Функционал по работе с конструктором бургера реализован полностью корректно - молодец!
- Соблюдены все стадии АПИ запроса, отраженные в экшенах (_REQUEST, _SUCCESS, _FAILED)
- Работа с redux реализована отлично
- Правильно реализована работа с хуками useSelector, useDispatch
- Правильно реализована работа с библиотекой react-dnd - отлично!

**Но есть несколько замечаний:**

**Замечания которые необходимо будет исправить:**
- Необходимо установить библиотеку prop-types в зависимости package.json
- Стоит избавиться от неиспользуемого контекста IngredientContext (включая файл burger-context.js)
- Реализован скролл при навигации - отлично! Однако при клике на активный раздел, сперва происходит скролл окна браузера,
  а только затем скролл списка ингредиентов. Вот от скролла окна браузера целесообразнее избавиться

Так же стоит учесть комментарии, оставленные в коде.

**Места, в которых можно улучшить проект:**
- Перед отправкой работы на pull-реквест, стоит отчищать код от издержек процесса разработки, а именно убрать из кода:
  неиспользуемые импорты, неиспользуемые переменные или функции, console, debugger,
  даже закомментированный код (его очень много в работе).
- В пакетах package.json есть @reduxjs/toolkit, однако сторы redux (экшены и редьюсеры) описаны в классическом стиле,
  без использования слайсов (преимуществ данной библиотеки). О работе с данной библиотекой и быстром внедрении в проект,
  можно почитать здесь: https://redux-toolkit.js.org/tutorials/quick-start
- Стоит обращать внимание на предупреждения в терминале, в котором запущен проект.
  Это хорошие рекомендации, которых следовало бы придерживаться

Отвечаю на ваш вопрос: "в правильном ли направлении движусь и правильно ли понимаю, как нужно работать с redux" - правильно, критических замечаний по работе с redux в данной работе нет.

Пожалуйста, учтите выше описанные замечания и те, что оставлены коде вашей работы (особенно с пометкой "нужно исправить").
При повторной проверке, если все замечания будут исправлены, работа будет принята.

Удачного рефакторинга кода!

EldenLordCommander committed:
fix ошибок по ревью

- очистил код от неиспользуемых элементов, контекста и закомментированных блоков
- поправил скроо списка ингредиентов
- добавил библиотеку prop-types

EldenLordCommander commented:
Добрый день.
Поправил замечания по коду:
- установил prop-types
- избавился от контекста
- поправил скролл списка ингридиентов (немного поправил стили элемента, чтобы скролл не затрагивал окно браузера)
- поправил некоторые мелочи

Пока что я хочу остановится на обычном Redux - мне его проще пока понимать, и еще есть элементы, которые бы хотел для себя закрепить, получше понять. Использовать возможности @reduxjs/toolkit планирую позже, если будет время, когда буду уверен в пользовании обычным redux (хотя бы для себя хочу получше его понять, а потом перейти к toolkit)

vadmas commented:
Отлично, все критичные комментарии исправлены. Работу принимаю. У вас хорошая база и амбиции, удачи в дальнейшем обучении!
